src/rust/engine/fs/store/src/remote/reapi.rs

```diff
@@ -143,40 +149,80 @@ impl Provider {
 
     let mut client = self.byte_stream_client.as_ref().clone();
 
-    let chunk_size_bytes = self.chunk_size_bytes;
+    // we have to communicate the (first) error reading the underlying reader out of band
+    let error_occurred = Arc::new(parking_lot::Mutex::new(None));
+    let error_occurred_stream = error_occurred.clone();
 
-    let stream = futures::stream::unfold((0, false), move |(offset, has_sent_any)| {
-      if offset >= len && has_sent_any {
-        futures::future::ready(None)
-      } else {
-        let next_offset = min(offset + chunk_size_bytes, len);
-        let req = protos::gen::google::bytestream::WriteRequest {
+    let chunk_size_bytes = self.chunk_size_bytes;
+    let stream = async_stream::stream! {
+      if len == 0 {
+        // if the reader is empty, the ReaderStream gives no elements, but we have to write at least
+        // one.
+        yield protos::gen::google::bytestream::WriteRequest {
           resource_name: resource_name.clone(),
-          write_offset: offset as i64,
-          finish_write: next_offset == len,
-          // TODO(tonic): Explore using the unreleased `Bytes` support in Prost from:
-          // https://github.com/danburkert/prost/pull/341
-          data: bytes(offset..next_offset),
+          write_offset: 0,
+          finish_write: true,
+          data: Bytes::new(),
         };
-        futures::future::ready(Some((req, (next_offset, true))))
+      } else {
```

Comment:
This could be less indented if the `len == 0` case were handled as an early return (generally my preference for short "guard" cases like this).